File: Python/message_buffer.py
import queue
import random
import time

from message import Message


class MessageBuffer:
    def __init__(self, capacity: int):
        self._capacity = capacity
        self._buffer = queue.Queue(maxsize=capacity)

    @property
    def capacity(self) -> int:
        return self._capacity

    @property
    def is_empty(self) -> bool:
        return self._buffer.empty()

    @property
    def is_full(self) -> bool:
        return self._buffer.full()

    def put(self, message: Message) -> None:
        self._buffer.put(message)
        time.sleep(random.random() * 100 / 1000)

    def take(self) -> Message:
        message = self._buffer.get()
        time.sleep(random.random() * 100 / 1000)
        return message

File: C++/src/message_buffer.cpp
#include <message_buffer.hpp>

MessageBuffer::MessageBuffer(size_t maxSize) : maxSize(maxSize) {}

void MessageBuffer::offer(const Message &message)
{
  std::unique_lock<std::mutex> lock(mtx);
  notFull.wait(lock, [this]
               { return buffer.size() < maxSize; });
  buffer.push(message);
  notEmpty.notify_all();
}

Message MessageBuffer::poll()
{
  std::unique_lock<std::mutex> lock(mtx);
  notEmpty.wait(lock, [this]
                { return !buffer.empty(); });
  Message message = buffer.front();
  buffer.pop();
  notFull.notify_all();
  return message;
}

bool MessageBuffer::isEmpty() const
{
  return buffer.empty();
}

bool MessageBuffer::isFull() const
{
  return buffer.size() == maxSize;
}
